### DISCUSSION TITLE
Building Consensus Protocols in SQLAlchemy?

### DISCUSSION THREAD

**Discussion by kapilagrawal95** [sqlalchemy/sqlalchemy#7342] [Ideas]
Hi, I had this idea about building a consensus protocol in SQLAlchemy. I came across this article on using multiple databases with Binds in SQLAlchemy: https://flask-sqlalchemy.palletsprojects.com/en/2.x/binds/ . This got me thinking whether we can start thinking of building consensus algorithms on top of these multiple databases. So here's a mental model I have:

<img width="830" alt="image" src="https://user-images.githubusercontent.com/12952642/142261855-e6a54a71-28fe-4915-ba57-1ef96c20630f.png">

Do you think this would be a valuable addition to SQLAlchemy?

**Comment by zzzeek** [comment]
well there's the question of what this consensus refers towards, are you trying to select a write master in a cluster, or coordinate a distributed transaction, determine if a database cluster has quorum, etc. .  the big "PAXOS" box about could be there for any number of reasons.

for most issues like these there is usually some kind of middleware that is handling the coordination of database nodes, which makes this "answer" available in such a way that any number of client applications can talk to the "cluster" without being involved in these details, and it looks ilke your "PAXOS" box above is serving that need, but usually this service would have to be some kind of proxying service, or otherwise setting up some kind of single source of truth that the client applications refer towards.

if OTOH you tried to have client applications themselves make decisions about the database nodes in a cluster, like, is there quorum, who is the write master, etc., that's more problematic because all of those client applications would have to be talking to each other as well to have agreement.

more commonly, schemes that embed into the "bind selection" aspect of the SQLAlchemy Session for the purposes of talking to HA clusters are much simpler than this, going on the assumption that the job of providing an HA cluster is handled outside the application server, such as using something like Galera, and that we just need to choose nodes efficiently, where we would use things like round robin.    But even then, you're better off using proxying middleware like HAProxy or PGBouncer for that anyway.   I will add that I did spend some time a few years back coming up with a way to embed round-robin or similar style connection load balancing at the connection pool level, though I abandoned it because HAProxy and other tools do this a lot better.

so it's not too clear what we want to do having a paxos layer bundled within client applications, but even if we do, that may be a very worthwhile project but there's no need that such a thing would need to be an "addition" to SQLAlchemy, that's not something the SQLAlchemy project would maintain directly.  It would be instead delivered as a third party extension.     I would also note that the Session.get_bind() idea would likely benefit from providing an event hook in order to override it, rather than relying upon subclassing the Session itself.  that's about the most I'd want to change SQLAlchemy itself in this area.
